gtsam/slam/SmartProjectionRigFactor.h

```diff
@@ -0,0 +1,366 @@
+    return cameraRig_;
+  }
+
+  /// return the calibration object
+  inline FastVector<size_t> cameraIds() const {
+    return cameraIds_;
+  }
+
+  /**
+   * print
+   * @param s optional string naming the factor
+   * @param keyFormatter optional formatter useful for printing Symbols
+   */
+  void print(const std::string& s = "", const KeyFormatter& keyFormatter =
+                 DefaultKeyFormatter) const override {
+    std::cout << s << "SmartProjectionRigFactor: \n ";
+    for (size_t i = 0; i < nonUniqueKeys_.size(); i++) {
+      std::cout << "-- Measurement nr " << i << std::endl;
+      std::cout << "key: " << keyFormatter(nonUniqueKeys_[i]) << std::endl;
+      std::cout << "cameraId: " << cameraIds_[i] << std::endl;
+      cameraRig_[ cameraIds_[i] ].print("camera in rig:\n");
+    }
+    Base::print("", keyFormatter);
+  }
+
+  /// equals
+  bool equals(const NonlinearFactor& p, double tol = 1e-9) const override {
+    const This *e = dynamic_cast<const This*>(&p);
+    return e && Base::equals(p, tol)
+    && nonUniqueKeys_ == e->nonUniqueKeys()
+    && cameraRig_.equals(e->cameraRig())
+    && std::equal(cameraIds_.begin(), cameraIds_.end(), e->cameraIds().begin());
+  }
+
+  /**
+   * Collect all cameras involved in this factor
+   * @param values Values structure which must contain body poses corresponding
+   * to keys involved in this factor
+   * @return vector of cameras
+   */
+  typename Base::Cameras cameras(const Values& values) const override {
+    typename Base::Cameras cameras;
+    cameras.reserve(nonUniqueKeys_.size());  // preallocate
+    for (size_t i = 0; i < nonUniqueKeys_.size(); i++) {
+      const Pose3 world_P_sensor_i = values.at<Pose3>(nonUniqueKeys_[i])  // = world_P_body
+      * cameraRig_[cameraIds_[i]].pose();  // = body_P_cam_i
+      cameras.emplace_back(
+          world_P_sensor_i,
+          make_shared<typename CAMERA::CalibrationType>(
+              cameraRig_[cameraIds_[i]].calibration()));
+    }
+    return cameras;
+  }
+
+  /**
+   * error calculates the error of the factor.
+   */
+  double error(const Values& values) const override {
+    if (this->active(values)) {
+      return this->totalReprojectionError(this->cameras(values));
+    } else {  // else of active flag
+      return 0.0;
+    }
+  }
+
+  /**
+   * Compute jacobian F, E and error vector at a given linearization point
+   * @param values Values structure which must contain camera poses
+   * corresponding to keys involved in this factor
+   * @return Return arguments are the camera jacobians Fs (including the jacobian with
+   * respect to both body poses we interpolate from), the point Jacobian E,
+   * and the error vector b. Note that the jacobians are computed for a given point.
+   */
+  void computeJacobiansWithTriangulatedPoint(typename Base::FBlocks& Fs,
+                                             Matrix& E, Vector& b,
+                                             const Cameras& cameras) const {
```

Comment:
Weird that another set of cameras is passed in here. I though we just had poses and the camera rig, OR cameras computed from them. Sanity-check?

Reply:
the cameras here include the body poses (from the values) and the sensor-to-body poses (fixed, from the rig). The function "cameras" is in charge of assembling them to recover the sensor poses in the global frame. Please unresolve if this is still of concern.